chia/wallet/wallet_singleton_store.py

```diff
@@ -0,0 +1,235 @@
+        self = cls()
+        self.db_wrapper = wrapper
+
+        async with self.db_wrapper.writer_maybe_transaction() as conn:
+            await conn.execute(
+                (
+                    "CREATE TABLE IF NOT EXISTS singletons("
+                    "coin_id blob PRIMARY KEY,"
+                    " coin text,"
+                    " singleton_id blob,"
+                    " wallet_id int,"
+                    " parent_coin_spend blob,"
+                    " inner_puzzle_hash blob,"
+                    " confirmed tinyint,"
+                    " confirmed_at_height int,"
+                    " spent_height int,"
+                    " lineage_proof blob,"
+                    " custom_data blob,"
+                    " generation int,"
+                    " timestamp int)"
+                )
+            )
+
+            await conn.execute("CREATE INDEX IF NOT EXISTS singleton_id_index ON singletons(singleton_id)")
+            await conn.execute("CREATE INDEX IF NOT EXISTS wallet_id_index ON singletons(wallet_id)")
+
+            await conn.execute(
+                "CREATE INDEX IF NOT EXISTS confirmed_at_height_index " "ON singletons(confirmed_at_height)"
+            )
+
+        return self
+
+    async def _clear_database(self) -> None:
+        async with self.db_wrapper.writer_maybe_transaction() as conn:
+            await (await conn.execute("DELETE FROM singletons")).close()
+
+    async def add_singleton_record(self, record: SingletonCoinRecord) -> None:
+        """
+        Store SingletonCoinRecord in DB
+        """
+        async with self.db_wrapper.writer_maybe_transaction() as conn:
+            columns = (
+                "coin_id, coin, singleton_id, wallet_id, parent_coin_spend, inner_puzzle_hash, "
+                "confirmed, confirmed_at_height, spent_height, lineage_proof, custom_data, "
+                "generation, timestamp"
+            )
+            await conn.execute(
+                f"INSERT or REPLACE INTO singletons ({columns}) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
+                (
+                    record.name(),
+                    json.dumps(record.coin.to_json_dict()),
+                    record.singleton_id,
+                    record.wallet_id,
+                    bytes(record.parent_coin_spend),
+                    record.inner_puzzle_hash,
+                    int(record.confirmed),
+                    record.confirmed_at_height,
+                    record.spent_height,
+                    bytes(record.lineage_proof),
+                    record.custom_data,
+                    record.generation,
+                    record.timestamp,
+                ),
+            )
+
+    async def get_records_by_singleton_id(
+        self,
+        singleton_id: bytes32,
+        min_generation: Optional[uint32] = None,
+        max_generation: Optional[uint32] = None,
+        num_results: Optional[uint32] = None,
```

Comment:
It's a good idea, but the newly introduced problem of possibly introducing a bug because of unexpected behavior (truncated results) seems worse to me.